File: assign04/parallel_sort.c
/*parallel_sort.c: compiled with gcc parallel_sort.c -pthread. ran with ./a.out*/
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

void* sortsublists(void *args);
void* mergesublists(void *args);

struct mergeparam_s { //this struct is used to store struct variables for the third thread two merge the two sublists into the single sorted list.
	double *arrin; /*this represents the input  array*/
	double *arrout; /*this represents the output array*/
	int n;
	int midl;
};

struct sortparam_s { //this struct is used to store struct variables for the threads to sort the sublists
	double *array;
	int beggng;
	int procend;
};

int main(int argc, char *argv[]) {
	pthread_t thr1; //the first p thread
	
	pthread_t thr2; //the second pthread
	
	pthread_t thr3; // the third p thread
	
	struct sortparam_s p1, p2;
	
	struct mergeparam_s p3;
	
	int N, mid, i;
	
	double * array; /* array of type double */
	
	double * array2; /* Used for copy of 'array'*/
	
	double * mergedarr; /* For storing merged array */
	
	clock_t start, end;
	
	if (argc != 2) {
		printf("Usage:%s<N>\n", argv[0]);
			
		return -1;
	}
	
	array = (double*)malloc(N * sizeof(double));
	
	array2 = (double*)malloc(N * sizeof(double));
	
	mergedarr = (double*)malloc(N * sizeof(double));
	
	N = atoi(argv[1]);
	
	srand(time(NULL)); //seed the random number generator used by rand() function.
	
	/*printf("Array: ");*/
	for(i = 0; i < N; i++) {/* generate the random numbers from the rand() function. */
	array[i] = array2[i] = (double)rand() / RAND_MAX * 99.0 + 1.0; /* range 1.0 - 100.0 */
	/* printf("%f ", arr[i]); */
	}
	/* printf("\n"); */
	
	p1.beggng = 0;
	
	p1.procend = N / 2;
	
	p1.array = array;
	
	p2.beggng = N /2;
	
	p2.procend = N;
	
	p2.array = array;
	
	p3.midl = N / 2; 
	
	p3.n = N;
	
	p3.arrin = array;
	
	p3.arrout = mergedarr;
	
	start = clock(); /*initialize the start time; set to the number of clock ticks*/
	
	//Now I am going to start the threads to begin working on the sub arrays
	pthread_create(&thr1, NULL, sortsublists, &p1); //create the first thread for sorting the first sublist, p1.
	
	pthread_create(&thr2, NULL, sortsublists, &p2); //create the second thread for sorting the first sublist, p1.
	
	//Wait for the sorting to complete...
	pthread_join(thr1, NULL);
	
	pthread_join(thr2, NULL);
	
	//Now I start the thread used for merging the two sublists
	pthread_create(&thr3, NULL, mergesublists, &p3);
	
	//Now I wait for the merging of the threads to complete
	pthread_join(thr3, NULL);
	
	end = clock() - start; //Set the end time; equal to number of current clock ticks minus the number of previous ones
	
	printf("Sorting is done in %fms when two threads are utilized\n", end*1000.0f / CLOCKS_PER_SEC);
	
	p1.procend = N; //This is for running on the entire array.
	
	p1.array = array2; //The first array, 'array', is already partially sorted. Now we pass the original copy of the array.
	
	start = clock(); //reinitialize the start time, equal to # of clock ticks
	
	sortsublists(&p1);
	
	end = clock() - start; //end time = # of current clock ticks minus the starting time
	
	printf("Sorting is done in %fms when one thread is utilized\n", end * 1000.0f / CLOCKS_PER_SEC);
	
	/*Now I am going to free the memory of the sorted sublists*/
	free(array);
	
	free(array2);
	
	free(mergedarr);
	
	return 0;
}

void* mergesublists(void *args) {
	int i, j, k;
	
	double *arrin, *arrout;
	
	struct mergeparam_s* proc = (struct mergeparam_s*) args;
	
	/* printf("Merge %d : ", proc->n); */
	
	i = k = 0; //initializing k and i to zero.
	
	j = proc->midl;
	
	arrin = proc -> arrin;
	
	arrout = proc -> arrout;
	
	while(i < proc ->midl && j < proc -> n){
		if(arrin[i] < arrin[j]) 
			arrout[k++] = arrin[i++]; 
		else
			arrout[k++] = arrin[j++];
	}
	
	while(i < proc -> midl) 
		arrout[k++] = arrin[i++];
	while(j < proc ->n)
		arrout[k++] = arrin[j++];
	/* for(i = 0; i < proc->n; i++)
     printf("%f ", arrout[i]);
	printf("\n"); */
	return NULL;
}

void* sortsublists(void *args) {
	int i, j, minim;
	
	double temp;
	
	double *array;
	
	struct sortparam_s* proc = (struct sortparam_s*) args;
	
	array = proc -> array;
	
	// printf("sort %d to %d: ", proc->beggng, proc->procend); 
	
	for(i = proc -> beggng; i < proc -> procend; i++) {
		minim = i;
		
		for(j = i + 1; j < proc -> procend; j++) {
			if(array[j] < array[minim])
				minim = j;
		}
		temp = array[i]; //this array hold the value of array for ith iteration from beggining of process until end of process.
		
		array[i] = array[minim];
		
		array [minim] = temp;
	}
	/*for(i = proc->beggng; i < proc->procend; i++)
    printf("%f ", array[i]);
	printf("\nEnd of the sort\n"); */	
	
	return NULL;
}

File: assign04/ParallelSort.java
//Java implementation of multithreaded app.: ParallelSort.java. Compiled with javac ParallelSort.java. Ran as 'java ParallelSort 5000' etc..
import java.util.*;

	class ThreadMerger extends Thread {
		private double [] arrin; //input arrayin
		
		private double [] arrout; //output array 
		
		private int midl; //used as dividing point into the sublists.
		
		public ThreadMerger(double [] i, double [] o, int m) {
			arrin = i;
			
			arrout = o;
			
			midl = m;
		}
		
		public void runThread() {
			int i, j, k;
			
			//System.out.println("Merged sublists " + arrin.length + " : ");
			
			i = k = 0; //initialize i and k to zero.
			
			j = midl;
			
			while(i < midl && j < arrin.length) { //use java string length method to return length of arrin's length
				if(arrin[i] < arrin[j])
					arrout[k++] = arrin[i++];
				else
					arrout[k++] = arrin[j++];
			}
			
			while(i < midl) 
				arrout[k++] = arrin[i++];
			while(j < arrin.length)
				arrout[k++] = arrin[j++];
			/*for(i = 0; i < arrin.length; i++)
            System.out.print(" " + arrout[i]);
            System.out.println(""); */
		}
	};
	
	class ThreadSorter extends Thread {
		private double [] arr; //the array which is to be sorted
		
		private int arrbeg, arrend;//starting and ending points in the array to be sorted.
		
		public ThreadSorter(double [] a, int b, int n) { //class constructor for ThreadSorter class, initialize the array objects.
			arr = a;
			
			arrbeg = b;
			
			arrend = n;
		}
		
		public void runThread() {
			double temp; //temporary variable 
			
			int i, j, minim;
			
			arr = arr;
			
			//System.out.println("Sort array from  " + arrbeg + " to " + arrend);
			
			for(i = arrbeg; i <arrend; i++) { //from the beggining to the end of the array
				minim = i;
				for(j = i + 1; j < arrend; j++) {
					if(arr[j] < arr[minim])
						minim = j;
				}
				temp = arr[i]; //the temp variable is set to the ith position of the sublist
				
				arr[i] = arr[minim]; //the ith position of the sublist is set to the 'minimum' element of the sublist
				
				arr[minim] = temp; //the 'minimum' element of the sublist is then set to the temporary variable.
			}
			/*for(i = arrbeg; i < arrend; i++)
            System.out.print(" " + arr[i]);
            System.out.println("\nEnd of sort"); */
		}
	};
	
	public class ParallelSort {
		
		public static void main(String []arg) {
			int i, N, midl;
			
			double [] arr; //first array
			
			double [] arrcop; //Used for a copy of array
			
			double [] mergedarr; //stores the third merged array.
			
			long start, end;
			
			if(arg.length != 1) {
				System.out.println("Usage: java ParallelSort <N>");
				
				return;
			}
			
			N = Integer.parseInt(arg[0]); //user Integer.parseInt java method to convert the user's input string to an integer and store in 'N'.
			
			arr = new double[N];
			
			arrcop = new double[N];
			
			mergedarr = new double[N];
			
			Random rand = new Random(System.currentTimeMillis());
			
			//System.out.println ("Array is: ");
			
			for(i = 0; i < N; i++) { /* generate  sequence of random numbers */
            arr[i] = arrcop[i] = rand.nextDouble() * 99.0 + 1.0; /* range is from 1.0 - 100.0 */
			
            //System.out.print(" "+ arr[i]);
			}
			//System.out.println("");

			//Now I am going to create the threads, first the two sorting threads, then the 'merging' thread 
			ThreadSorter sort1 = new ThreadSorter(arr, 0, N/2);
			
			ThreadSorter sort2 = new ThreadSorter(arr, N/2, arr.length);
			
			ThreadMerger merge1 = new ThreadMerger(arr, mergedarr, N/2);
			
			start = System.nanoTime(); //The start time; initialize start to value of most precise 'system' timer.
			
			/* Now I am going to 'start' the threads to begin their work on the sublists(sub-arrays).*/
			sort1.start();
			
			sort2.start();
			
			//Waiting for the sublist sorting to complete...
			try {
				sort1.join();
				
				sort2.join();
			} catch (Exception e) {
				e.printStackTrace(); //used to print Java.lang.Throwable class and other details, in this case of the exception 'e'.
			}
			
			/*Now I am going to start the thread for merging...*/
			merge1.start();
			/*Wait for the merger thread to finish...*/
		try {
			merge1.join();
		} catch (Exception e) {
				e.printStackTrace();
		}
		
		end = System.nanoTime() - start; //This will represent the ending time.
		
		System.out.println("Sorting is done within " + end/1000000.0f + "ms when two threads are utilized");
		
		ThreadSorter completeSort = new ThreadSorter(arrcop, 0, N); //For running along the entirety of the array.
		
		start = System.nanoTime(); //the start time..
		
		completeSort.start();
		
		try {
			completeSort.join();
		} catch (Exception e) {
			e.printStackTrace();
		}
		
		end = System.nanoTime() - start; //the end time...
		
		System.out.println("Sorting is done within " + end/1000000.0f + "ms when one single thread is utilized");
	}
} //compile with javac ParallelSort.java. run as 'java ParallelSort 5000' etc..
